### PR TITLE
materialized: bundle console in image

### PR DESCRIPTION
**Pull request by benesch**
To make it even easier to get started with Materialize, this commit bundles a recent copy of the console into the emulator image (i.e., the materialize/materialized image).

This reuses the console packaging that we already have in place to bundle the console with the self-managed distribution.

We'll need to remember to periodically bump the console image, but it shouldn't be too bad--the only time we *must* bump the image is when we make a backwards incompatible change to a feature the console is depending on, which doesn't happen very often.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by SangJunBak** [comment]
@kay-kim to test, you can just run the emulator install command
```
 docker run -d -p 127.0.0.1:6874:6874 -p 127.0.0.1:6875:6875 -p 127.0.0.1:6876:6876 materialize/materialized:v0.128.0-dev.0--pr.g8d9686caf7d4fb2c7b0249a591cfd07bccd08b80
```
and you'd use the tag from the buildkite/CI build. This is how you get it:

https://github.com/user-attachments/assets/2653cc9d-ea5e-4b9d-93a1-5ad455da55ba

**Comment by SangJunBak** [comment]
@benesch I appended some commits that'd need your review but https://github.com/MaterializeInc/materialize/pull/30822/commits/85cbcedb1174ec8cc62bcd8428bba409dd7808db isn't working and idk why. We need to disable RBAC such that our create UI is enabled and the query history doesn't show a "you don't have privileges" screen. Your changes look good to me btw

**Comment by benesch** [comment]
> @benesch I appended some commits that'd need your review but [85cbced](https://github.com/MaterializeInc/materialize/commit/85cbcedb1174ec8cc62bcd8428bba409dd7808db) isn't working and idk why. We need to disable RBAC such that our create UI is enabled and the query history doesn't show a "you don't have privileges" screen. Your changes look good to me btw

Ah, what's the source for setting that `MZ_DISABLE_AUTHENTICATION` env var? I can't find any references to it in the code. What we do in orchestratord to disable RBAC is to set `enable_rbac_checks=off`:

https://github.com/MaterializeInc/materialize/blame/9211a02ef16e52378ab9531e1e878664ec52a989/src/orchestratord/src/controller/materialize/environmentd.rs#L862-L864

We need to be careful here though, as this image is used in many tests, many of which test RBAC. So we can't just disable RBAC here. cc @def- for advice. 

Certainly the simplest thing to do would be to just live with the console being a bit broken in the emulator for a bit, and file a ticket about fixing it in the future.

**Comment by SangJunBak** [comment]
>Ah, what's the source for setting that MZ_DISABLE_AUTHENTICATION env var? I can't find any references to it in the code. What we do in orchestratord to disable RBAC is to set enable_rbac_checks=off:

Ah I must've been looking at https://github.com/benesch/materialize/blob/console-in-emulator/src/orchestratord/src/controller/materialize.rs#L56 which was a helm chart operator argument and not an environment variable 🙈 My bad!

>We need to be careful here though, as this image is used in many tests, many of which test RBAC. So we can't just disable RBAC here. cc @def- for advice.

Ah that's a good catch. Out of curiosity, is `materialized` used by the helm chart? I noticed I didn't see any services/pods for it when spinning it up in `kind` so I kinda assumed it was for just the emulator / local builds.

>Certainly the simplest thing to do would be to just live with the console being a bit broken in the emulator for a bit, and file a ticket about fixing it in the future.

An issue with this tho is a lot of the Console's functionality is broken when the user isn't a super user / RBAC isn't disabled. From what I can tell:
- You can't create clusters and other objects via the UI because we hide it
- You can't delete clusters and objects from the UI because we hide it
- We show a "you lack permissions" screen for the query history

We hide most of the UI because our ownership check is broken. This is bc. HTTP requests auto-register each query with user `anonymous_http_user`, even if we pass it a `user` header. Furthermore, the Console doesn't have ownership of any objects you create in external clients like psql.

**Comment by benesch** [comment]
> Ah that's a good catch. Out of curiosity, is `materialized` used by the helm chart? I noticed I didn't see any services/pods for it when spinning it up in `kind` so I kinda assumed it was for just the emulator / local builds.

It's not! It's used for emulator, local builds, and many of the CI tests in this repository.



> We hide most of the UI because our ownership check is broken. This is bc. HTTP requests auto-register each query with user `anonymous_http_user`, even if we pass it a `user` header. Furthermore, the Console doesn't have ownership of any objects you create in external clients like psql.

Huh. This seems fixable. Seems like the HTTP request should use the auth header if no Frontegg auth is configured—this more closely matches what we do for the SQL interface. Something like this should do it:

```diff
diff --git a/src/environmentd/src/http.rs b/src/environmentd/src/http.rs
index 077d8e1faf..e2f486c460 100644
--- a/src/environmentd/src/http.rs
+++ b/src/environmentd/src/http.rs
@@ -717,8 +717,16 @@ async fn http_auth(
         (TlsMode::Require, ConnProtocol::Https { .. }) => {}
     }
     let creds = match frontegg {
-        // If no Frontegg authentication, use the default HTTP user.
-        None => Credentials::DefaultUser,
+        None => {
+            // If no Frontegg authentication, use whatever is in the HTTP auth
+            // header (without checking the password), or fall back to the
+            // default user.
+            if let Some(basic) = req.headers().typed_get::<Authorization<Basic>>() {
+                Credentials::User(basic.username().to_string())
+            } else {
+                Credentials::DefaultUser
+            }
+        }
         Some(_) => {
             if let Some(basic) = req.headers().typed_get::<Authorization<Basic>>() {
                 Credentials::Password {
```

If you have a spare few minutes this week @SangJunBak maybe you could try that patch out and see if you can get it working? 

IIUC, this would make the ownership check start working, which would solve most of the problem. The query history would still be disabled because the `materialize` role wouldn't have access to it, but if that's the only problem, that feels like something we can definitely punt on.

File: doc/user/content/get-started/install-materialize-emulator.md [hunk 1]
```diff
@@ -37,7 +37,7 @@ not suitable for production deployments</redb>.
 | **Use case isolation**               | ❌ | ✅ |
 | **Fault tolerance**                  | ❌ | ✅ |
 | **Horizontal scalability**           | ❌ | ✅ |
-| **GUI**      | ❌ | ✅ Materialize Console |
+| **GUI**      | ✅ Materialize Console  | ✅ Materialize Console |
```
Comment (morsapaes) [inline doc/user/content/get-started/install-materialize-emulator.md:40]:
```suggestion
| **GUI (Materialize Console)**      | ✅ | ✅ |
```

File: doc/user/content/get-started/install-materialize-emulator.md [hunk 2]
```diff
@@ -59,21 +59,24 @@ not suitable for production deployments</redb>.
...
 
    ```sh
-   docker run -d -p 127.0.0.1:6875:6875 -p 127.0.0.1:6876:6876 materialize/materialized:{{< version >}}
+   docker run -d -p 127.0.0.1:6874:6874 -p 127.0.0.1:6875:6875 -p 127.0.0.1:6876:6876 materialize/materialized:{{< version >}}
    ```
 
    When running locally:
 
    - The Docker container binds exclusively to localhost, for security reasons.
+   - The web console is available on port `6874`.
```
Comment (morsapaes) [inline doc/user/content/get-started/install-materialize-emulator.md:68]:
```suggestion
   - The [Materialize console](/console/) is available on port `6874`.
```

File: doc/user/content/get-started/install-materialize-emulator.md [hunk 3]
```diff
@@ -59,21 +59,24 @@ not suitable for production deployments</redb>.
...
    - The SQL interface is available on port `6875`.
    - Logs are available via `docker logs <container-id>`.
    - A default user `materialize` is created.
    - A default database `materialize` is created.
 
 1. <a name="materialize-emulator-connect-client"></a>
 
-   Connect to the Materialize Emulator using your preferred SQL client, using
-   the following connection field values:
+   Open the Materialize Console in your browser at `http://localhost:6874`.
```
Comment (morsapaes) [inline doc/user/content/get-started/install-materialize-emulator.md:76]:
```suggestion
   Open the Materialize console in your browser at `http://localhost:6874`.
```

File: doc/user/content/get-started/quickstart.md [hunk 4]
```diff
@@ -52,15 +52,14 @@ Emulator](/get-started/install-materialize-emulator/) to test locally. However,
...
+## Step 0. Open the SQL Shell
 
 - If you have a Materialize account, navigate to the [Materialize
   Console](https://console.materialize.com/) and sign in. By default, you should
   be in the SQL Shell. If you're already signed in, you can access the SQL Shell in the left-hand menu.
 
-- If you are using the Materialize Emulator, [connect to the Materialize
-  Emulator](/get-started/install-materialize-emulator/#materialize-emulator-connect-client)
-  using your preferred SQL client.
+- If you are using the Materialize Emulator, open the Materialize Console at
```
Comment (morsapaes) [inline doc/user/content/get-started/quickstart.md:61]:
```suggestion
- If you are using the Materialize Emulator, open the Materialize console at
```

File: doc/user/content/get-started/quickstart.md [hunk 5]
```diff
@@ -457,67 +429,21 @@ data arrives (either from the known_flippers table or the flip_activities view).
...
-   ) WITH (snapshot = false)) TO STDOUT;
-   ```
-
-   The optional [`WITH (snapshot = false)`
-   option](/sql/subscribe/#with-options) indicates that the command displays
-   only the new flippers that come in after the start of the `SUBSCRIBE`
-   operation, and not the flippers in the view at the start of the operation.
-
-   1. Open another SQL client and connect to your Materialize.
+1. In the Materialize Console quickstart page, enter an id (for example `450`)
```
Comment (morsapaes) [inline doc/user/content/get-started/quickstart.md:432]:
```suggestion
1. In the Materialize console quickstart page, enter an id (for example `450`)
```

File: doc/user/content/get-started/install-materialize-emulator.md (hunk 2 above)
Comment (morsapaes) [reply doc/user/content/get-started/install-materialize-emulator.md:68]: @kay-kim, we seem to use both "Materialize Console" and "Materialize console" in the documentation. Can we decide on one variation and stick to it?

Comment (kay-kim) [reply doc/user/content/get-started/install-materialize-emulator.md:68]: Yup -- going to go with Materialize Console (capitalizing both).

**Comment by kay-kim** [comment]
Hey @SangJunBak  -- I know you're adding a patch for RBAC ...  just adding this here for a quick check when the patch is up.
Screenshot 2024-12-16 at 11 36 52 AM

**Comment by SangJunBak** [comment]
Added a new commit [Add enabled cluster sizes and disable rbac for materialized](https://github.com/MaterializeInc/materialize/pull/30822/commits/81a73d491afb85653f703b1e9af9ec217bac641b) such that we disable RBAC by default anyways since CI will most likely override it. This means the commit to allow to recognize username/password auth when there's no frontegg isn't useful, but we still want it since `anonymous_http_user` queries are logged otherwise.

**Review by SangJunBak — APPROVED** [review]
Approving all commits/changes that aren't my own

**Review by kay-kim — APPROVED** [review]
Also good with the docs part of these changes (ran through quickstart using the console).  Q: When/how is the new emulator released? Because until it's released, the docs would be a little bit out of sync with the current emulator, yes?

**Comment by SangJunBak** [comment]
@kay-kim Judging by https://hub.docker.com/r/materialize/materialized/tags?page=3, we tag 'latest' as the last tag (so currently latest is v0.127.1)

Screenshot 2024-12-16 at 11 26 30 PM

and the docs for installing the emulator don't really specify a version. 

Screenshot 2024-12-16 at 11 28 12 PM

so we can either:
1. Merge this PR w/ the doc commits and code commits, and allow the docs to be a little out of sync until v0.128 releases
2. Separate the doc commits from the code commits and merge the doc commits once v0.128 releases

What do you think though?

**Comment by benesch** [comment]
> Also good with the docs part of these changes (ran through quickstart using the console). Q: When/how is the new emulator released? Because until it's released, the docs would be a little bit out of sync with the current emulator, yes?

Yes, that's right. The emulator is released the moment the release is cut, so 10pm ET on any Thursday we're cutting a release. So the simplest thing to do would be to merge this PR a few minutes/hours before that release cut occurs. 

Unfortunately the next release doesn't get cut until Jan 3. So I'm kind of inclined to say that this Friday we should merge this PR, then immediately backport and cut a release of v0.127.1 not for release to production but _just_ to have a release of the emulator that includes the console. Then the docs will correctly point to that version of the emulator, which will include the console, and the docs will tell everyone to use the console for the quickstart, and all will be well.

@SangJunBak would you be up for coordinating that on Friday?

**Comment by benesch** [comment]
Oh, sorry, @SangJunBak, I didn't see your comment before I posted mine. The docs _do_ specify an emulator version though, no? 

That `{{< version >}}` tag turns into the latest release tag directly, e.g. right now it's `v0.126` in the published docs:

image

**Comment by SangJunBak** [comment]
>That {{< version >}} tag turns into the latest release tag directly, e.g. right now it's v0.126 in the published docs:

💀 

> @SangJunBak would you be up for coordinating that on Friday?

@benesch Sounds like a plan!

**Comment by benesch** [comment]
> @benesch Sounds like a plan! Lemme know if my changes look good and I can merge!

Ah sorry yes, they look great! Go for it whenever.

**Comment by benesch** [comment]
(I'll just handle it now, per @morsapaes's request.)

**Comment by benesch** [comment]
Merged and backported to v0.127.2 so that this is available in the emulator immediately.

**Comment by SangJunBak** [comment]
Thanks for pushing this through @benesch!

**Comment by benesch** [comment]
Thanks for making it work in the first place!

**Pull request merged**